packages/connect-popup/src/index.tsx

```diff
@@ -218,70 +141,215 @@ const handleMessage = (
+                break;
+            case 'Method_PermissionsNotGranted':
+            case 'Method_Cancel':
+                // User canceled process, close popup.
+                window.close();
+                break;
+            default:
+                fail({
+                    type: 'error',
+                    detail: 'response-event-error',
+                    message: data.payload?.error || 'Unknown error',
+                });
+                analytics.report({
+                    type: EventType.ViewChangeError,
+                    payload: { code: data.payload?.code || 'Code missing' },
+                });
+        }
+    }
+};
+
+/**
+ * receive initial configuration from npm package
+ */
+const handleInitMessage = (event: MessageEvent<PopupEvent | IFrameLogRequest>) => {
+    const { data } = event;
+    if (!data) return;
+
+    if (data.type === IFRAME.LOG) {
+        log.debug(data.payload);
+        return;
+    }
+
+    // This is message from the window.opener
+    if (data.type === POPUP.INIT) {
+        init(escapeHtml(data.payload));
+        window.removeEventListener('message', handleInitMessage);
+    }
+};
+
+const handleMessageInIframeMode = (
+    event: MessageEvent<
+        | PopupEvent
+        | UiEvent
+        | (Omit<MethodResponseMessage, 'payload'> & { payload?: { error: string; code?: string } })
+    >,
+) => {
+    const { data } = event;
+
+    if (!data) return;
+
+    if (disposed) return;
+
+    log.debug('handleMessage', data);
+    handleResponseEvent(data);
+
+    // This is message from the window.opener
+    if (data.type === UI_REQUEST.IFRAME_FAILURE) {
+        fail({
+            type: 'error',
+            detail: 'iframe-failure',
+        });
 
-    if (!payload.systemInfo) {
-        payload.systemInfo = getSystemInfo(config.supportedBrowsers);
+        return;
     }
 
-    const isBrowserSupported = await view.initBrowserView(payload.systemInfo);
-    if (!isBrowserSupported) {
+    // ignore messages from origin other then MessagePort (iframe)
+    const isMessagePort =
+        event.target instanceof MessagePort ||
+        (typeof BroadcastChannel !== 'undefined' && event.target instanceof BroadcastChannel);
+    if (!isMessagePort) return;
+
+    // catch first message from iframe
+    if (data.type === POPUP.HANDSHAKE) {
+        handshake(data, event.origin);
         return;
     }
 
+    const message = parseMessage(data);
+
+    analytics.report({ type: EventType.ViewChange, payload: { nextView: message.type } });
+
+    if (
+        message?.payload &&
+        typeof message.payload === 'object' &&
+        'analytics' in message.payload &&
+        message.payload.analytics
+    ) {
+        analytics.report(message.payload.analytics);
+    }
+
+    handleUIAffectingMessage(message);
+};
+
+// handle POPUP.INIT message from window.opener
+const init = async (payload: PopupInit['payload']) => {
+    log.debug('popup init', payload);
+
+    if (!payload) return;
```

Comment:
Isn't this repeating the same 4 lines bellow?

Reply:
yeah, right, I will remove this